### ISSUE TITLE
TCAV with TensorFlow Object Detection API models

### ISSUE THREAD

**Issue by emmanuelol** [tensorflow/tcav#113]
Hi, I run the repo without issues, and I was able to apply it to Keras models (image classification) successfully. Now I'm looking at how to apply it to Object Detection. But I'm pretty lost because the Object Detection API model is an SSDMetaArch file. If you use the *.pb file, don't provide access to backbone layers directly, and these layers are on Keras. Hence I can apply the Keras wrapper neither use one that follows a graph such as the examples provided in this repository.
Any idea or suggestion on where I can start?
Quickly add what I'm able to do:

Get the Keras classification backbone of the Object detector model
Get the bottleneck of this model
Get the input tensor of this model
I cannot pull the loss function from the model, but I can write one.
What not being able to do;

Get the prediction layer
not sure if I need to get the prediction layer by individual bounding boxes
I really appreciate any help you can provide.
P.S. if more details or info is needed from my side. please let me know

**Comment by emmanuelol** [comment]
Hi, me again, quickly want to add what I'm able to do:

- Get the Keras classification backbone of the Object detector model
- Get the bottleneck of this model
- Get the input tensor of this model
- I cannot pull the loss function from the model, but I can write one.

What not being able to do;

- Get the prediction layer
- not sure if I need to get the prediction layer by individual bounding boxes

**Comment by emmanuelol** [comment]
Hi, to update my recent findings and issues.

 Findings: 
- tracing back to the objectdetection class, I found out how to get the prediction layer of the SSD.
- Also, I successfully recreate the loss function for this layer that works in the same fashion as KerasWrapper
Issues:

- I copy the KerasWrapper class to build a new KerasWrapperSSD (the SSD layer are Keras layers).
- This class is also on model.py of the TCAV repository
- I load the model as I did to found the model's layers, no error in that, except that the model seems that is no loading the backbone or any other layer. Hence I cannot extract the bottlenecks.
```python
  def import_keras_model(self, saved_path):
    """Loads keras model, fetching bottlenecks, inputs and outputs."""
    self.ends = {}
    PATH_TO_CFG=saved_path + "/pipeline.config"
    PATH_TO_CKPT = saved_path + "/checkpoint"
    configs = config_util.get_configs_from_pipeline_file(PATH_TO_CFG)
    model_config = configs['model']
    detection_model = model_builder.build(model_config=model_config, is_training=False)
    ckpt = tf.compat.v2.train.Checkpoint(model=detection_model)
    ckpt.restore(os.path.join(PATH_TO_CKPT, 'ckpt-0')).expect_partial()
    self.model = detection_model
    print(self.model)
    print(self.model.feature_extractor.classification_backbone)
    self.get_bottleneck_tensors()
    self.get_inputs_and_outputs_and_ends()
    self.get_image_shape()
```

- If I the load model in this fashion on other script works, but as a class of TCAV, the model is just a simple shell
I don't know what is missing or what I'm doing wrong. I really, really appreciate any help.
Bests

**Comment by emmanuelol** [comment]
Hi, I try different setups of the wrapper but remain the same result. It seems that I cannot load an Object Detection model in this repository. Any help or guidance on why or how to solve this issue is welcome. I really appreciate any help you can provide.

**Comment by jameswex** [comment]
@BeenKim do you have any examples of people using TCAV for TF/Keras object detection models?

**Comment by BeenKim** [comment]
Hi @emmanuelol , I'm a bit unclear on which part of TCAV that doesn't seem to work. You mentioned " Hence I cannot extract the bottlenecks." - what's the error msg you are getting?

I'll close this issue for now since it's been a while (perhaps you've solved it by now), but feel free to reopen if needed.
Been
